sstables/sstable_directory.cc

```diff
@@ -65,15 +65,21 @@ std::unique_ptr<sstable_directory::components_lister>
 sstable_directory::make_components_lister() {
     return std::visit(overloaded_functor {
         [this] (const data_dictionary::storage_options::local& loc) mutable -> std::unique_ptr<sstable_directory::components_lister> {
-            return std::make_unique<sstable_directory::filesystem_components_lister>(make_path(_table_dir, _state));
+            if (loc.dir.empty()) {
+                on_internal_error(dirlog, "Non-table local storage options");
```

Comment:
Instead of all this checking, why not make it so that options can *only* be created with dir/prefix?

Reply:
Well, first note that storage options are referenced by `table`s and by `keyspace_metadata`s. And those options for different tables are also different, because each table has its own dir/prefix in storage. Also note, that the storage_options::s3::prefix is not always a prefix in a bucket, it can be a key in system.sstables table, unfortunately. I'm going to fix this further.

So why not have dir/prefix here all the time? The reason is in keyspace_metadata options. There s3::prefix doesn't make any sense -- system.sstables doesn't contain records related to keyspaces. The local::dir might make sense, because keyspace has its directory on disk, but I left it empty for symmetry with s3 options, and for this particular sanity check -- make_components_lister() (as well as make_storage(), that has similar check) are to be called only for table's storage options, if called for keyspace options it indicated a runtime error that's better be captured  early.